conductor/src/action.hpp

```diff
@@ -0,0 +1,198 @@
+
+#ifndef COCAINE_CONDUCTOR_ACTION_HPP
+#define COCAINE_CONDUCTOR_ACTION_HPP
+
+
+namespace cocaine { namespace isolate { namespace conductor {
+
+namespace action {
+
+struct cancellation_t:
+    public api::cancellation_t
+{
+    typedef std::shared_ptr<action_t> parent_ptr;
+    typedef std::weak_ptr<action_t> parent_wptr;
+
+    parent_wptr m_parent;
+
+    cancellation_t(parent_ptr parent);
+
+    virtual void cancel();
+};
+
+
+class action_t:
+    public enable_shared_from_this<action_t>
+{
+    // enum states {
+    //     st_pending = 0,
+    //     st_done,
+    //     st_cancelled
+    // };
+
+protected:
+
+    uint64_t             m_request_id;
+    shared_ptr<client_t> m_parent;
+
+    shared_ptr<io::basic_dispatch_t> m_dispatch;
+
+public:
+
+    struct states {
+        const static int st_pending = 0;
+        const static int st_done = 1;
+        const static int st_cancelled = 2;
+    };
+
+    std::atomic<int>     m_state;
+
+
+    typedef session<asio::ip::tcp> session_type;
+
+    action_t(shared_ptr<client_t> client):
+        m_state(states::st_pending),
+        m_request_id(++client->max_request_id),
+        m_parent(client)
+    {}
+
+    template<class Derived>
+    shared_ptr<Derived>
+    shared_self(){
+        return std::static_pointer_cast<Derived>(shared_from_this());
```

Comment:
this is really BAD idea. You need dynamic cast here, however even presence of dynamic cast indicates problems.